src/NumericalAlgorithms/LinearSolver/ConjugateGradient/InitializeElement.hpp

```diff
@@ -34,55 +35,89 @@ struct InitializeResidual;
     // residual monitor
     Parallel::contribute_to_reduction<
         cg_detail::InitializeResidual<ParallelComponent>>(
         Parallel::ReductionData<
             Parallel::ReductionDatum<double, funcl::Plus<>>>{
-            inner_product(r, r)},
+            inner_product(residual, residual)},
         Parallel::get_parallel_component<ParallelComponent>(cache)[array_index],
         Parallel::get_parallel_component<ResidualMonitor<Metavariables>>(
             cache));
 
-    return db::create_from<db::RemoveTags<>, simple_tags, compute_tags>(
-        std::move(box),
-        // We have not started iterating yet, so the _next_ iteration ID is
-        // zero. We initialize the current iteration id such that it advances to
-        // zero by the increment operator.
-        std::numeric_limits<size_t>::max(), size_t{0}, std::move(r),
-        db::item_type<LinearSolver::Tags::HasConverged>{});
+    using compute_tags = db::AddComputeTags<
+        ::Tags::NextCompute<LinearSolver::Tags::IterationId>>;
+    return std::make_tuple(
+        ::Initialization::merge_into_databox<
+            InitializeElement,
+            db::AddSimpleTags<LinearSolver::Tags::IterationId, residual_tag,
+                              LinearSolver::Tags::HasConverged>,
+            compute_tags, MergePolicy>(
+            std::move(box),
+            // We have not started iterating yet, so we initialize the current
+            // iteration ID such that the _next_ iteration ID is zero.
+            std::numeric_limits<size_t>::max(), std::move(residual),
+            db::item_type<LinearSolver::Tags::HasConverged>{}),
+        // Terminate algorithm for now. The reduction will be broadcast to the
+        // next action which is responsible for restarting the algorithm.
+        true);
+  }
+};
+
+struct InitializeHasConverged {
+  template <typename ParallelComponent, typename DataBox,
+            typename Metavariables, typename ArrayIndex,
+            Requires<db::tag_is_retrievable_v<LinearSolver::Tags::HasConverged,
+                                              DataBox>> = nullptr>
+  static void apply(DataBox& box,
+                    Parallel::ConstGlobalCache<Metavariables>& cache,
+                    const ArrayIndex& array_index,
+                    const db::item_type<LinearSolver::Tags::HasConverged>&
+                        has_converged) noexcept {
+    db::mutate<LinearSolver::Tags::HasConverged>(
+        make_not_null(&box), [&has_converged](
+                                 const gsl::not_null<db::item_type<
+                                     LinearSolver::Tags::HasConverged>*>
+                                     local_has_converged) noexcept {
+          *local_has_converged = has_converged;
+        });
+
+    // Proceed with algorithm
+    Parallel::get_parallel_component<ParallelComponent>(cache)[array_index]
+        .perform_algorithm(true);
```

Comment:
I think it would be a little bit more clear if this instead of here was executed at the end of the `InitializeResidual`-Action that is calling `InitializeHasConverged`.

Reply:
That's on another parallel component - the residual monitor singleton. It receives the reduction and broadcasts it back to the element array, which then proceeds with the algorithm.